x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointAction.java

```diff
@@ -0,0 +1,214 @@
+        ) {
+            this.task = task;
+            this.listener = listener;
+            this.nodesAndShards = nodesAndShards;
+            this.nodes = clusterState.nodes();
+            this.localNodeId = clusterService.localNode().getId();
+        }
+
+        public void start() {
+            GroupedActionListener<GetCheckpointNodeAction.Response> groupedListener = new GroupedActionListener<>(
+                ActionListener.wrap(responses -> {
+                    // the final list should be ordered by key
+                    Map<String, long[]> checkpointsByIndexReduced = new TreeMap<>();
+
+                    // merge the node responses
+                    for (GetCheckpointNodeAction.Response response : responses) {
+                        response.getCheckpoints().forEach((index, checkpoint) -> {
+                            if (checkpointsByIndexReduced.containsKey(index)) {
+                                long[] shardCheckpoints = checkpointsByIndexReduced.get(index);
+                                for (int i = 0; i < checkpoint.length; ++i) {
+                                    shardCheckpoints[i] = Math.max(shardCheckpoints[i], checkpoint[i]);
+                                }
+                            } else {
+                                checkpointsByIndexReduced.put(index, checkpoint);
+                            }
+                        });
+                    }
+
+                    listener.onResponse(new Response(checkpointsByIndexReduced));
+                }, listener::onFailure),
+                nodesAndShards.size()
+            );
+
+            for (Entry<String, Set<ShardId>> oneNodeAndItsShards : nodesAndShards.entrySet()) {
+                if (localNodeId.equals(oneNodeAndItsShards.getKey())) {
+                    TransportGetCheckpointNodeAction.getGlobalCheckpoints(indicesService, oneNodeAndItsShards.getValue(), groupedListener);
+                    continue;
+                }
+
+                GetCheckpointNodeAction.Request nodeCheckpointsRequest = new GetCheckpointNodeAction.Request(
+                    oneNodeAndItsShards.getValue()
+                );
+                nodeCheckpointsRequest.setParentTask(clusterService.localNode().getId(), task.getId());
+                DiscoveryNode node = nodes.get(oneNodeAndItsShards.getKey());
+
+                // paranoia: this should not be possible using the same cluster state
+                if (node == null) {
+                    listener.onFailure(
+                        new UnavailableShardsException(
+                            oneNodeAndItsShards.getValue().iterator().next(),
+                            "Node not found for [{}] shards",
+                            oneNodeAndItsShards.getValue().size()
+                        )
+                    );
+                    return;
+                }
+
+                logger.trace("get checkpoints from node {}", node);
+                transportService.sendRequest(
```

Comment:
```suggestion
                transportService.sendChildRequest(
```

Sending a child request parents the request to this task